scripts/metalk8s_cli/mixins/salt.py

```diff
@@ -0,0 +1,119 @@
+        except subprocess.CalledProcessError as exc:
+            raise CommandError(
+                'Failed to execute "{}" [retcode {}]:\n{}'.format(
+                    ' '.join(full_cmd), exc.returncode, exc.output
+                )
+            )
+
+    def run_salt_master(self, command):
+        # TODO: add cache for the salt-master container ID
+        full_cmd = [
+            'crictl', 'exec', self.get_salt_master_container()
+        ]
+        master_cmd = ['salt-run', '--force-color'] + self._build_args(command)
+        full_cmd.extend(master_cmd)
+
+        try:
+            return subprocess.check_output(
+                # FIXME: pillar args... need to escape?
+                ' '.join(full_cmd), shell=True,
+                stderr=subprocess.STDOUT,
+            )
+        except subprocess.CalledProcessError as exc:
+            raise CommandError(
+                'Failed to run "{}" on Salt master:\n{}'.format(
+                    ' '.join(master_cmd), exc.output
+                )
+            )
+
+    def get_from_pillar(self, pillar_key):
+        return self._get_json_from_minion('pillar.get {}'.format(pillar_key))
+
+    def get_grain(self, grain_name, local=False):
+        command = 'grains.get {}'.format(grain_name)
+        return self._get_json_from_minion(command, local=local)
+
+    def render_jinja(self, template):
+        command = 'slsutil.renderer string="{}"'.format(template)
+        return self._get_json_from_minion(command)
+
+    @property
+    def minion_id(self):
+        # FIXME: add cache
+        return self.get_grain('id', local=True)
+
+    def get_saltenv(self):
+        return "metalk8s-{}".format(
+            self.render_jinja('{{ pillar.metalk8s.nodes[grains.id].version }}')
+        )
+
+    @staticmethod
+    def get_salt_master_container():
+        try:
+            output = subprocess.check_output([
+                'crictl', 'ps', '-q',
+                '--label', 'io.kubernetes.container.name=salt-master'
+            ])
+        except subprocess.CalledProcessError as exc:
+            raise CommandError(
+                'Failed to find salt-master container: {}'.format(exc)
+            )
+
+        return output.strip().decode('utf-8')
+
+    def check_role(self, role):
+        """Ensure this command runs on a Node with this role."""
+        minion_roles = self.render_jinja(
+            '{{ pillar.metalk8s.nodes[grains.id].roles | tojson }}'
+        )
+
+        if role not in minion_roles:
+            raise CommandInitError(
+                "Can only run this command on a '{}' Node.".format(role)
+            )
+
+    def _build_args(self, cmd):
+        if isinstance(cmd, six.string_types):
+            return shlex.split(cmd)
+        return cmd
+
+    def _get_json_from_minion(self, command, **kwargs):
+        out_string = self.run_salt_minion(
+            command, outputter='json', color=False, **kwargs
+        )
+        return json.loads(out_string)["local"]
```

Comment:
Care here since you do `stderr=subprocess.STDOUT` in subprocess call if for what ever reason salt log something to stderr you will likely not be able to parse this as json.

Reply:
Indeed, will change. I think I'll remove `check_output` and use `run` instead, that will allow more control for callers.

Reply:
`subprocess.run` was introduced in Python 3.5, dunno if it's "old enough" to be used in this context.

Reply:
Again... OK so I'll do another monkey-patch of this `subprocess.run` as we do for `argparse`, the ugliness of these should encourage us in only using Python 3 :)

Reply:
I mean, it's our custom script so maybe we can use Python 3, but even then I'm not sure CentOS 7 comes with Python 3.5+ (it's Python 3.4, no?)

Reply:
There's no Python 3 pre-installed on CentOS 7. There's 3.6 in `base`, though most of the usual dependencies (e.g. `six` or `PyYAML`) are not, so we'd need to include them in our archive, and somehow install them before execution of the bootstrap script (since it's intended to run through this Python CLI).

I think it could be feasible, but would cost more efforts for us. So let's go with something simpler for now :)

Reply:
Mmh, monkey-patch is not that easy, there's quite a lot of changes in the `run` API actually, when compared to the capabilities of `subprocess` in 2.7... So I'll ignore `run` and only expose `stderr` in case of error.

Reply:
`run` provides a bit more flexibility, so I made a _poor man's_ variant of it using `Popen.communicate`. It lacks `KeyboardInterrupt` or timeout management though, but should be fine for a first version.

Reply:
What's the deal with `KeyboardInterrupt`?

For timeout indeed, there is nothing in Python 2…

Reply:
So, `run` in Python 3 relies on using `Popen` as a contextmanager, which performs clean-up of open streams and handles `KeyboardInterrupt` errors when doing so. I just ignored the clean-up of stdin (which may raise other errors), as it's not in our use-cases for now. See [`subprocess.Popen.__exit__`](https://github.com/python/cpython/blob/master/Lib/subprocess.py#L998-L1024)